lib/model/model.go

```diff
@@ -1291,6 +1343,56 @@ func (m *model) ClusterConfig(conn protocol.Connection, cm protocol.ClusterConfi
 	return nil
 }
 
+func (m *model) ensureIndexHandler(conn protocol.Connection) *indexHandlerRegistry {
+	deviceID := conn.DeviceID()
+	connID := conn.ConnectionID()
+
+	m.pmut.Lock()
+	defer m.pmut.Unlock()
+
+	indexHandlerRegistry, ok := m.indexHandlers[deviceID]
+	if ok && indexHandlerRegistry.conn.ConnectionID() == connID {
+		// This is an existing and proper index handler for this connection.
+		return indexHandlerRegistry
+	}
+
+	if ok {
+		// A handler exists, but it's for another connection than the one we
+		// now got a ClusterConfig on. This should be unusual as it means
+		// the other side has decided to start using a new primary
+		// connection but we haven't seen it close yet. Ideally it will
+		// close shortly by itself...
+		l.Infof("Abandoning old index handler for %s (%s) in favour of %s", deviceID.Short(), indexHandlerRegistry.conn.ConnectionID(), connID)
+	}
+
+	// Create a new index handler for this device.
+	indexHandlerRegistry = newIndexHandlerRegistry(conn, m.deviceDownloads[deviceID], m.closed[connID], m.Supervisor, m.evLogger)
+	for id, fcfg := range m.folderCfgs {
+		l.Infoln("Registering folder", id, "for", deviceID.Short())
+		indexHandlerRegistry.RegisterFolderState(fcfg, m.folderFiles[id], m.folderRunners[id])
+	}
+	m.indexHandlers[deviceID] = indexHandlerRegistry
+
+	return indexHandlerRegistry
+}
+
+func (m *model) getIndexHandler(conn protocol.Connection) (*indexHandlerRegistry, bool) {
+	deviceID := conn.DeviceID()
+	connID := conn.ConnectionID()
+
+	m.pmut.RLock()
+	defer m.pmut.RUnlock()
```

Comment:
Similar concern as in https://github.com/syncthing/syncthing/pull/8918#discussion_r1285198517: I fear that moving the lock inside a helper function increases the chance of this being missed in future changes and thus potentially introducing lock issues.

Reply:
(fixd)